### PR TITLE
refactor(client): Extract a form prefill model.

### LINKED ISSUE #1992: Uses of `prefillEmail` & `prefillPassword` have expanded past their original intentions and are causing problems
Link to #1973 

`prefillEmail` and `prefillPassword` were originally added to allow a user to go from signIn->signUp->back or signIn->TOS->back w/o losing the data they entered. `prefillEmail` ended up being relied upon to provide the notion of a "previously signed in account", which is what User/Accounts has taken over. The two do not always join cleanly, and where they don't, an unexpected user can be signed in.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
@zaach or @vladikoff - could you check this one out?

* No longer use Session as a dumping ground for prefill information.

fixes #1992

### CODE REVIEW

File: app/scripts/views/confirm_reset_password.js [hunk 1]
```diff
@@ -188,7 +186,6 @@ function (Cocktail, _, $, ConfirmView, BaseView, Template, p, Session, Constants
       // users will be redirected back to the RP, Sync users will be
       // taken to the Sync controlled completion page.
       Session.clear();
-      Session.set('prefillEmail', self._email);
```
Comment (shane-tomlinson) [inline app/scripts/views/confirm_reset_password.js]: formPrefill is not cleared, so no need re-populate it.

File: app/scripts/lib/session.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ define([
   var NAMESPACE = '__fxa_session';
 
   // and should not be saved to sessionStorage
-  var DO_NOT_PERSIST = ['prefillPassword', 'prefillYear', 'error'];
+  var DO_NOT_PERSIST = ['error'];
```
Comment (zaach) [inline app/scripts/lib/session.js]: yass

File: app/tests/spec/views/confirm_reset_password.js [hunk 3]
```diff
@@ -17,13 +17,14 @@ define([
...
   'models/user',
+  'models/form-prefill',
   '../../mocks/router',
   '../../mocks/window',
   '../../lib/helpers'
 ],
 function (chai, sinon, p, AuthErrors, View, Session, Metrics, EphemeralMessages,
-      InterTabChannel, Storage, FxaClient, Relier, Broker, User, RouterMock,
-      WindowMock, TestHelpers) {
+      InterTabChannel, Storage, FxaClient, Relier, Broker, User, FormPrefill,
```
Comment (zaach) [inline app/tests/spec/views/confirm_reset_password.js]: It doesn't look like `FormPrefill` is used in this file?

**Comment by zaach** [comment]
Coveralls shows the router module has dropped in coverage but I'm not sure what caused it. From the change in coverage of the `watchAnchor` method, maybe a link that was triggered before is no longer being triggered.

**Comment by shane-tomlinson** [comment]
> Coveralls shows the router module has dropped in coverage but I'm not sure what caused it. From the change in coverage of the watchAnchor method, maybe a link that was triggered before is no longer being triggered.

I'm going to open a separate PR with a boost to the router.js coverage. The diff is larger than I want to add on to this one.

**Comment by shane-tomlinson** [comment]
Opened #2148 to boost test coverage of router.js

**Comment by zaach** [comment]
So close to killing `Session`!!!11 :+1:

**Pull request merged**